irc/handlers.go

```diff
@@ -3097,6 +3097,203 @@ func markReadHandler(server *Server, client *Client, msg ircmsg.Message, rb *Res
 	return
 }
 
+// METADATA <target> <subcommand> [<and so on>...]
+func metadataHandler(server *Server, client *Client, msg ircmsg.Message, rb *ResponseBuffer) (exiting bool) {
+	if !server.Config().Metadata.Enabled {
+		return unknownCommandHandler(server, client, msg, rb)
+	}
+
+	originalTarget := msg.Params[0]
+	target := originalTarget
+	subcommand := strings.ToLower(msg.Params[1])
+
+	invalidTarget := func() {
+		rb.Add(nil, server.name, "FAIL", "METADATA", "INVALID_TARGET", originalTarget, client.t("Invalid metadata target"))
+	}
+	noKeyPerms := func(key string) {
+		rb.Add(nil, server.name, "FAIL", "METADATA", "KEY_NO_PERMISSION", originalTarget, key, client.t("You do not have permission to perform this action"))
+	}
+
+	targetIsSelf := target == "*"
+	targetIsANick := server.clients.Get(target) != nil
+	targetIsAChan := server.channels.Get(target) != nil
+
+	if !targetIsSelf && !targetIsAChan && !targetIsANick {
+		invalidTarget()
+		return
+	}
+	if targetIsSelf {
+		target = client.nick
+	}
+	if !metadataCanISeeThisTarget(client, target) {
+		invalidTarget()
+		return
+	}
+
+	needsKey := subcommand == "set" || subcommand == "get" || subcommand == "sub" || subcommand == "unsub"
+	if needsKey && len(msg.Params) <= 2 {
+		// no idea if this is the right thing to send here
+		rb.Add(nil, server.name, ERR_NEEDMOREPARAMS, client.Nick(), msg.Command, client.t("Not enough parameters"))
+		return
+	}
+
+	switch subcommand {
+	case "set":
+		key := msg.Params[2]
+		if metadataKeyIsEvil(key) {
```

Comment:
Add length validation for key + value (I think "combined length at most 350 bytes" is fine for now)